drivers/pcf857x/pcf857x.c

```diff
@@ -0,0 +1,470 @@
+static int _write(const pcf857x_t *dev, pcf857x_data_t data);
+
+#if IS_USED(MODULE_PCF857X_IRQ)
+
+/* interrutp service routine for IRQs */
+static void _irq_isr(void *arg);
+
+/* declaration of IRQ handler function */
+static void _irq_handler(event_t *event);
+
+/* internal update function */
+static void _update_state(pcf857x_t* dev);
+
+#endif /* MODULE_PCF857X_IRQ */
+
+int pcf857x_init(pcf857x_t *dev, const pcf857x_params_t *params)
+{
+    /* some parameter sanity checks */
+    assert(dev != NULL);
+    assert(params != NULL);
+    assert(params->exp < PCF857X_EXP_MAX);
+#if IS_USED(MODULE_PCF857X_IRQ)
+    assert(gpio_is_valid(params->int_pin));
+#endif
+
+    DEBUG_DEV("params=%p", dev, params);
+
+    /* init device data structure */
+    dev->params = *params;
+
+    switch (params->exp) {
+#if IS_USED(MODULE_PCF8574)
+        /**< PCF8574 8 bit I/O expander used */
+        case PCF857X_EXP_PCF8574: dev->pin_num = PCF8574_GPIO_PIN_NUM;
+                                  dev->params.addr += PCF8574_BASE_ADDR;
+                                  break;
+#endif
+#if IS_USED(MODULE_PCF8574A)
+        /**< PCF8574A 8 bit I/O expander */
+        case PCF857X_EXP_PCF8574A: dev->pin_num = PCF8574A_GPIO_PIN_NUM;
+                                   dev->params.addr += PCF8574A_BASE_ADDR;
+                                   break;
+#endif
+#if IS_USED(MODULE_PCF8575)
+        /**< PCF8575 16 bit I/O expander */
+        case PCF857X_EXP_PCF8575: dev->pin_num = PCF8575_GPIO_PIN_NUM;
+                                  dev->params.addr += PCF8575_BASE_ADDR;
+                                  break;
+#endif
+        default: return -PCF857X_ERROR_INV_EXP;
+    }
+
+#if IS_USED(MODULE_PCF857X_IRQ)
+    /* initialize the IRQ event object used for delaying interrupts */
+    dev->irq_event.event.handler = _irq_handler;
+    dev->irq_event.dev = dev;
+
+    for (unsigned i = 0; i < dev->pin_num; i++) {
+        dev->isr[i].cb = NULL;
+        dev->isr[i].arg = NULL;
+        dev->enabled[i] = false;
+    }
+
+    /* initialize the interrupt pin */
+    if (gpio_init_int(dev->params.int_pin,
+                      GPIO_IN_PU, GPIO_FALLING, _irq_isr, (void*)dev)) {
+        return -PCF857X_ERROR_INT_PIN;
+
+    }
+#endif /* MODULE_PCF857X_IRQ */
+
+    int res = PCF857X_OK;
+
+    _acquire(dev);
+
+    /* write 1 to all pins to switch them to INPUTS pulled up to HIGH */
+    dev->out = ~0;
+    res |= _write(dev, dev->out);
+
+    /* initial read all pins */
+    res |= _read(dev, &dev->in);
+
+    /* set all pin modes to INPUT and set internal output data to 1 (HIGH) */
+    dev->modes = ~0;
+
+    _release(dev);
+
+    return PCF857X_OK;
+}
+
+int pcf857x_gpio_init(pcf857x_t *dev, gpio_t pin, gpio_mode_t mode)
+{
+    /* some parameter sanity checks */
+    assert(dev != NULL);
+    assert(pin < dev->pin_num);
+
+    DEBUG_DEV("pin=%u mode=%u", dev, pin, mode);
+
+    switch (mode) {
+#ifndef MCU_CC2538
```

Comment:
To fix compilation errors due to this enumeration type :worried: https://github.com/RIOT-OS/RIOT/blob/fea525f01921abebb195ec8895d67d528f21aa96/cpu/cc2538/include/periph_cpu.h#L150-L158 where `GPIO_OD` and `GPIO_OD_PU` are declared as undefined. I didn't take a look at the implementation to see if these values are used directly as register values. But it wouldn't make any difference, because the special case would still have to be considered.

Reply:
But is `GPIO_OD` the same as `GPIO_IN`? Otherwise you can just let this case be hanled by `default` and not support it.

Reply:
A bit of reordering to handle everything that is not `GPIO_OUT` as default. The question is whether not supported `GPIO_IN_OD` should lead to an assert or simply a failure.

Reply:
As long as `gpio_init` returns an error code, you can do so too